### ISSUE TITLE
Duplicates NEO blocks when saving "Homepage" section

### ISSUE THREAD

**Issue by Osze** [spicywebau/craft-neo#156]
My NEO block in the entry "Homepage" gets duplicated 9! times when saving the section - "Homepage" (single)

I've made a short video explaining. 

https://youtu.be/lcW89xrKjwU

**Comment by Osze** [comment]
Created another Single Section "Test" and same issue. Happens only to NEO Blocks.

**Comment by ttempleton** [comment]
Hi @Osze,

There were duplicating block issues (actually extra block structures mistakenly being created) but they should have been resolved in v2.0.1... I've repeated the process as shown in your video and have as yet been unable to reproduce this.  I'll try to look through the plugin code and find any places that fields/blocks could possibly be updated and saved without the block structure also being updated and saved.  In the meantime, can I ask, since you've reported the issue with single entries; are you also experiencing this with channel or structure entries, or are they working normally?

**Comment by Osze** [comment]
Hi ttempleton, 

I only get the issue on singles.
Both Craft and NEO are using the latest version. 

If it helps I can give you access to the system. My name is Oscar Carlsson in the Craft Slack channel. Or add me on skype - oscar_carlsson84

Regards
Oscar

**Comment by ttempleton** [comment]
The offer is much appreciated; for the time being I'm going to see if there's any setting or combination of settings that could cause this, but if I come up empty then I'll be in touch.

**Comment by Osze** [comment]
I've been trying to find a pattern or something to understand the bug better. But without luck. 

For example:
I created a new Single. 
Added my NEO Block to it and saved some content. 
Went back to the Single and saved it. 

It all looked as it should. 

Clear:ed my cache and re-saved the Single Section again. Then the duplicated neo blocked appeard. 

hm...

**Comment by i-hardy** [comment]
I'm seeing this too, latest version of Neo and Craft 3.0.30. Every block in the page is duplicated and I can't actually open it in the CP.

**Comment by Osze** [comment]
@ttempleton did you have time to look at this issue? 

I'm hopefully launching the new site tomorrow and would feel much better if this was solved :)

**Comment by ttempleton** [comment]
Sorry for the delay @Osze and @i-hardy.  We've just released v2.0.2 which includes a clean-up of duplicated block structures.  If you update Neo and then re-save the affected sections, the appearance of duplicated blocks should be gone.  Can both of you please confirm whether this has resolved the situation for you?

**Comment by Osze** [comment]
Hi Ttempleton, 

I've upgraded NEO to v.2.0.2.

but re-savning entries doesn't work at all now. Throws me this error:

`
PDOException: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction in /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Command.php:1258
Stack trace:
#0 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Command.php(1258): PDOStatement->execute()
#1 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Command.php(1075): yii\db\Command->internalExecute('UPDATE `craft_e...')
#2 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/ActiveRecord.php(281): yii\db\Command->execute()
#3 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/BaseActiveRecord.php(802): yii\db\ActiveRecord::updateAll(Array, Array)
#4 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/ActiveRecord.php(625): yii\db\BaseActiveRecord->updateInternal(NULL)
#5 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/BaseActiveRecord.php(673): yii\db\ActiveRecord->update(false, NULL)
#6 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Elements.php(438): yii\db\BaseActiveRecord->save(false)
#7 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Sections.php(1290): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#8 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Sections.php(635): craft\services\Sections->_onSaveSingle(Object(craft\models\Section), false, Array)
#9 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/controllers/SectionsController.php(186): craft\services\Sections->saveSection(Object(craft\models\Section))
#10 [internal function]: craft\controllers\SectionsController->actionSaveSection()
#11 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#12 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#13 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Controller.php(104): yii\base\Controller->runAction('save-section', Array)
#14 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('save-section', Array)
#15 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(282): yii\base\Module->runAction('sections/save-s...', Array)
#16 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(542): craft\web\Application->runAction('sections/save-s...', Array)
#17 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(266): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#18 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#19 /srv/users/craft/apps/mybetting-in/public/index.php(21): yii\base\Application->run()
#20 {main}

Next yii\db\Exception: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction
The SQL being executed was: UPDATE `craft_elements` SET `fieldLayoutId`=3128, `enabled`=1, `archived`=0, `dateUpdated`='2018-11-19 08:45:20' WHERE `id`=2 in /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Schema.php:664
Stack trace:
#0 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Command.php(1263): yii\db\Schema->convertException(Object(PDOException), 'UPDATE `craft_e...')
#1 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/Command.php(1075): yii\db\Command->internalExecute('UPDATE `craft_e...')
#2 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/ActiveRecord.php(281): yii\db\Command->execute()
#3 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/BaseActiveRecord.php(802): yii\db\ActiveRecord::updateAll(Array, Array)
#4 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/ActiveRecord.php(625): yii\db\BaseActiveRecord->updateInternal(NULL)
#5 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/db/BaseActiveRecord.php(673): yii\db\ActiveRecord->update(false, NULL)
#6 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Elements.php(438): yii\db\BaseActiveRecord->save(false)
#7 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Sections.php(1290): craft\services\Elements->saveElement(Object(craft\elements\Entry))
#8 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/services/Sections.php(635): craft\services\Sections->_onSaveSingle(Object(craft\models\Section), false, Array)
#9 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/controllers/SectionsController.php(186): craft\services\Sections->saveSection(Object(craft\models\Section))
#10 [internal function]: craft\controllers\SectionsController->actionSaveSection()
#11 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#12 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#13 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Controller.php(104): yii\base\Controller->runAction('save-section', Array)
#14 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('save-section', Array)
#15 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(282): yii\base\Module->runAction('sections/save-s...', Array)
#16 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(542): craft\web\Application->runAction('sections/save-s...', Array)
#17 /srv/users/craft/apps/mybetting-in/craft/vendor/craftcms/cms/src/web/Application.php(266): craft\web\Application->_processActionRequest(Object(craft\web\Request))
#18 /srv/users/craft/apps/mybetting-in/craft/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#19 /srv/users/craft/apps/mybetting-in/public/index.php(21): yii\base\Application->run()
#20 {main}
Additional Information:
Array
(
    [0] => HY000
    [1] => 1205
    [2] => Lock wait timeout exceeded; try restarting transaction
)

Copy Stacktrace Search Stackoverflow Search Google Exception
Database Exception – yii\db\Exception
SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction
The SQL being executed was: UPDATE `craft_elements` SET `fieldLayoutId`=3128, `enabled`=1, `archived`=0, `dateUpdated`='2018-11-19 08:45:20' WHERE `id`=2
Error Info: Array
(
    [0] => HY000
    [1] => 1205
    [2] => Lock wait timeout exceeded; try restarting transaction
)`

**Comment by Osze** [comment]
Yeah, same issue when saving the entry directly. 

Tried it on 2 different craft installations and same problem. Endless loading until General error: 1205 Lock wait timeout exceeded;.

**Comment by Osze** [comment]
I have two installations with the same setup and same plugins. 

On both the mysql locks the table when saving the Hompage section and sometimes the Homepage entry:
```
SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction
The SQL being executed was: UPDATE craft_elements SET fieldLayoutId=3128, enabled=1, archived=0, dateUpdated='2018-11-20 02:54:41' WHERE id=2
```

I've tried to get Craft support helping me without any success. 

I have another installation that is almost the same, but not updated to the latest NEO 2.0.2. 
On that site the Homepage entry and Homepage section is able to save. No locking problems. 
...but then the neo blocks get duplicated (earlier issue) when saving the section instead.  

This really starts to get annoying. 

I'm running a m5.large on Amazon configured by Serverpilot, so the system is definitely good enough.

**Comment by Osze** [comment]
- Using MySQL 5.7.24 on the system where I get lock wait timeout. 
- MySQL 5.5.62 on the system where neo blocks get duplicated. Don't dare to update to NEO 2.0.2 because of the locking issue on the other server.

**Comment by Osze** [comment]
Hi ttempleton, you are right. I don't think it is a NEO issue anymore. 

I think it might have something with Blitz Cache. Just typically that Blitz also made an update at the same time you fixed #1 bug in NEO.
Sorry for taking your time on this!

**Comment by ttempleton** [comment]
No worries @Osze -- if it's a Blitz Cache issue, hopefully that gets sorted out soon.

I'll close this issue now, but if it does still turn out that Neo is involved, let us know and we'll reopen; and @i-hardy if you continue to experience issues with Neo after updating to v2.0.2, let us know.
